icu4c/source/i18n/dtptngen.cpp

```diff
@@ -654,6 +654,23 @@ void DateTimePatternGenerator::getAllowedHourFormats(const Locale &locale, UErro
 
     int32_t* allowedFormats = getAllowedHourFormatsLangCountry(language, country, status);
 
+    // We need to check if there is an hour cycle on locale
+    char buffer[8];
+    int32_t count = locale.getKeywordValue("hours", buffer, sizeof(buffer), status);
+
+    char16_t defaultHourFormatCharFromLocale = 0;
```

Comment:
Nit: Why not just set to `fDefaultHourFormatChar` here?  You can make it default to 0.